tests/app/questionnaire/test_answer_store_updater.py

```diff
@@ -0,0 +1,506 @@
+        ]
+
+        form = MagicMock()
+        form.serialise.return_value = answers
+
+        self.answer_store_updater.save_answers(form)
+
+        # answers should be passed straight through as Answer objects
+        answer_calls = list(map(mock.call, answers))
+        assert answer_calls in self.answer_store.add_or_update.call_args_list
+
+        self.answer_store.filter.return_value = iter(i[0][0].__dict__ for i in self.answer_store.add_or_update.call_args_list)
+
+        self.answer_store_updater.remove_empty_household_members()
+
+        remove_answer_calls = [call(answer_ids=['first-name', 'middle-names', 'last-name'], answer_instance=0),
+                               call(answer_ids=['first-name', 'middle-names', 'last-name'], answer_instance=1)]
+
+        # both instances of the answer should be removed
+        assert remove_answer_calls in self.answer_store.remove.call_args_list
+        assert self.answer_store.remove.call_count == 2
+
+    def test_remove_empty_household_members_values_entered_are_stored(self):
+        self.location.block_id = 'household-composition'
+        self.schema.get_group_dependencies.return_value = None
+        self.schema.get_answer_ids_for_block.return_value = ['first-name', 'middle-names', 'last-name']
+
+        answered = [
+            Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='first-name',
+                answer_instance=0,
+                value='Joe'
+            ), Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='middle-names',
+                answer_instance=0,
+                value=''
+            ), Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='last-name',
+                answer_instance=0,
+                value='Bloggs'
+            )
+        ]
+
+        unanswered = [
+            Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='first-name',
+                answer_instance=1,
+                value=''
+            ), Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='middle-names',
+                answer_instance=1,
+                value=''
+            ), Answer(
+                group_instance=0,
+                group_instance_id='group-0',
+                answer_id='last-name',
+                answer_instance=1,
+                value=''
+            )
+        ]
+
+        answers = []
+        answers.extend(answered)
+        answers.extend(unanswered)
+
+        form = MagicMock()
+        form.serialise.return_value = answers
+
+        self.answer_store_updater.save_answers(form)
+
+        self.answer_store.filter.return_value = iter(i[0][0].__dict__ for i in self.answer_store.add_or_update.call_args_list)
```

Comment:
🙀

Reply:
👀

Reply:
We shouldn't need to do it this way, we  just need a mock for the filter method to return the answers i.e. we can remove the `save_answers` and `add_or_update stuff`.

Reply:
Have updated the unit test to match this.